### PR TITLE
runtime-rs: Add basic boilerplate for remote hypervisor

### PR DESCRIPTION
**Pull request by Chasing1020**
This patch adds the support of the basic boilerplate of the remote hypervisor type.
- Generate remote hypervisor ttrpc to support peer pod VMs
- Enable annotation from oci-spec
- Add Remote hypervisor template config
- Use a Unix domain socket specified in the config file

Fixes: #10350

### CODE REVIEW

**Comment by Apokleos** [comment]
Hey @Chasing1020 Thx for your contribution! 
To better track changes, I suggest that create an issue for this PR(maybe it should be first created and then pull this request, but doesn't matter).  And then associate this PR with the issue.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/remote/inner.rs:74]: Comments needed to clarify the reason of using `10_0000_0000`

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/remote/inner.rs:82]: Directly using unwrap will cause panic in some conditions. If so, will it be expected ?

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/remote/inner.rs:98]: Could you please abstract the Client::connect(xxx) into a dedicated method to handle such similar works ?

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/remote/inner.rs:107]: useless comment needs be removed

File: src/runtime-rs/crates/hypervisor/src/remote/mod.rs
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/remote/mod.rs:1]: Copyright needed at the head of this new file
```
// Copyright 2024 Kata Contributors
//
// SPDX-License-Identifier: Apache-2.0
//
```

File: src/libs/kata-types/src/annotations/cri_containerd.rs
Comment (Apokleos) [inline src/libs/kata-types/src/annotations/cri_containerd.rs:16]: comments needed here to explain them.

File: src/libs/kata-types/src/config/default.rs
Comment (Apokleos) [inline src/libs/kata-types/src/config/default.rs:104]: Comments needed here about 600, users may not know exactly which one: 600s or 600ms or others ?

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/remote/inner.rs:72]: what's about `fs::File::open(netns_path).context("xxxxx")?;`

**Review by Apokleos — COMMENTED** [review]
Hi @Chasing1020 Good job! some initial comments left.

**Comment by Chasing1020** [comment]
Thanks a lot for so detailed comments! I will follow up on these comments and update the code shortly.

File: src/libs/kata-types/src/config/hypervisor/remote.rs
Comment (pmores) [inline src/libs/kata-types/src/config/hypervisor/remote.rs:73]: I believe the purpose of `validate()` is to perform a final validation of config's overall coherence after possible adjustments (like filling in defaults) have been made in `adjust_config()`.  I don't think it makes much sense to repeat here the checks that were already handled in `adjust_config()`.

File: src/runtime-rs/config/configuration-remote.toml.in
Comment (pmores) [inline src/runtime-rs/config/configuration-remote.toml.in:17]: Please document these settings similar to how existing setting are documented.  What is the remote hypervisor socket, who is the expected remote peer?  When is the time-out applied (connection setup? keep-alive?), which units is it given in?

File: src/libs/kata-types/src/config/hypervisor/remote.rs
Comment (pmores) [reply src/libs/kata-types/src/config/hypervisor/remote.rs:73]: For instance, if remote hypervisor doesn't support memory hotplugging (as documented elsewhere in this PR) it could be a good idea to verify `default_maxmemory` is not set, or equals to `default_memory`.

A similar check should be performed for `default_vcpus`/`default_maxvcpus`, considering that vcpu hotplugging is not supported either.

We should also ensure that no nvdimm usage is configured, and ideally check for all other remote hypervisor limitations (unsupported settings, setting where just a subset of possible values are valid for the remote hypervisor etc.).

File: src/runtime-rs/Makefile
Comment (pmores) [inline src/runtime-rs/Makefile:178]: `DEFSANDBOXCGROUPONLY` and `DEFSTATICRESOURCEMGMT` were [removed recently](https://github.com/kata-containers/kata-containers/pull/10052/commits/d11657a581243cef536b4c6cce6710bd22339bc9) - why are we adding them back?  @ananos can you please take a look?

Also, why is this PR touching dragonball and firecracker build?

File: src/runtime-rs/Makefile
Comment (pmores) [inline src/runtime-rs/Makefile:231]: Again, why is this PR modifying dragonball build?

File: src/runtime-rs/Makefile
Comment (Chasing1020) [reply src/runtime-rs/Makefile:474]: Thx, it looks like I mistakenly rebase the old submission (and sorry for the mistakes)

File: src/runtime-rs/crates/hypervisor/src/lib.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/lib.rs:132]: I'd hope that this won't be necessary and a better way can be found.

File: src/runtime-rs/crates/hypervisor/src/remote/mod.rs
Comment (pmores) [inline src/runtime-rs/crates/hypervisor/src/remote/mod.rs:194]: What is the intended purpose of this function?  If I'm not mistaken it collects a bunch of key/value pairs that it stores in `self.annotations` and uses in `prepare_vm()`.

However, most of the values are available to `prepare_vm()` already - default memory & vcpus are constants, machine type comes from config and namespace is passed to `prepare_vm()` as an argument.

That leaves sandbox name.  Is that the same thing as the id passed to `prepare_vm()` as another argument?  If not, could it be passed to the remote hv driver in some other way?  I think e.g. extending the `prepare_vm()` signature would be more acceptable than adding a whole new function to the `Hypervisor` trait that only serves a single implementation.  Also, more broadly, what is the purpose of sandbox name, how exactly is it used?  If it's just to identify a sandbox perhaps the sandbox id would serve the purpose even better?

File: src/runtime-rs/Makefile
Comment (pmores) [reply src/runtime-rs/Makefile:474]: Cool, thanks!

File: src/runtime-rs/crates/hypervisor/src/remote/protocols/hypervisor.proto
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/remote/protocols/hypervisor.proto:1]: append Copyright with Kata Containers Contributors

File: src/runtime-rs/crates/hypervisor/src/lib.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/lib.rs]: I have deleted this change.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs]: The timeout have changed to the default value. Makes it the same action as Go.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs]: I have added `new_ttrpc_client` method for `RemoteInner` to avoid them.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs]: Fixed by using `?`

File: src/runtime-rs/config/configuration-remote.toml.in
Comment (Chasing1020) [reply src/runtime-rs/config/configuration-remote.toml.in:19]: These are from the runtime of Go, and I want this configuration file to be compatible with the Go's.
And I have filled in the documents in my latest submission.

**Comment by Apokleos** [comment]
@Chasing1020 I think you should resolve conflicts.

File: src/runtime-rs/Makefile
Comment (gkurz) [reply src/runtime-rs/Makefile]: This is dropping the terminating `\n` which is bad.

File: src/runtime-rs/crates/hypervisor/Cargo.toml
Comment (gkurz) [inline src/runtime-rs/crates/hypervisor/Cargo.toml:77]: Please don't drop the terminating `\n` .

Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/Cargo.toml]: Fixed

File: src/libs/kata-types/src/annotations/cri_containerd.rs
Comment (gkurz) [inline src/libs/kata-types/src/annotations/cri_containerd.rs:18]: From https://pkg.go.dev/github.com/containerd/containerd/pkg/cri/annotations :
```suggestion
// SandboxNamespace is the name of the namespace of the sandbox (pod)
```

File: src/libs/kata-types/src/annotations/mod.rs
Comment (pmores) [inline src/libs/kata-types/src/annotations/mod.rs:324]: How are those annotation different from the existing `io.katacontainers.config.hypervisor.{machine_type,default_vcpus,default_memory}`?  Shouldn't we be using those?

File: src/libs/kata-types/src/config/default.rs
Comment (pmores) [inline src/libs/kata-types/src/config/default.rs:104]: What is timeouting here, which event starts the timer and which ends it?

File: src/runtime-rs/crates/hypervisor/src/remote/mod.rs
Comment (pmores) [inline src/runtime-rs/crates/hypervisor/src/remote/mod.rs:54]: Frankly, as a matter of style I'd prefer if you pass `annotations` to `RemoteInner::prepare_vm()` and keep this function's implementation trivial as is the convention.  This works too though...

**Comment by pmores** [comment]
@Chasing1020 very cool change avoiding storing annotations and instead producing & consuming them directly in `prepare_vm()`, thanks!

**Comment by gkurz** [comment]
Hi @Chasing1020 !

Thanks for reworking this PR. We're getting closer :smile:

Now I'd like you to do the following for each commit :
- add a description
- run `cargo fmt`

**Comment by Chasing1020** [comment]
@gkurz @pmores @Apokleos Thank you very much for your detailed code review and feedback. I will improve it soon.

File: src/libs/kata-types/src/config/hypervisor/mod.rs
Comment (gkurz) [inline src/libs/kata-types/src/config/hypervisor/mod.rs:1058]: It seems that this hunk and the corresponding users belong to some other commit, maybe the previous one.

File: src/runtime-rs/crates/hypervisor/build.rs
Comment (gkurz) [inline src/runtime-rs/crates/hypervisor/build.rs:8]: `clippy` is failing in `make check-runtime-rs` for me :

```
error: the borrowed expression implements the required traits
 --> crates/hypervisor/build.rs:8:17
  |
8 |         .inputs(&["src/remote/protocols/hypervisor.proto"])
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `["src/remote/protocols/hypervisor.proto"]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrows_for_generic_args
  = note: `-D clippy::needless-borrows-for-generic-args` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(clippy::needless_borrows_for_generic_args)]`

error: could not compile `hypervisor` (build script) due to previous error
```

Using rust 1.75.

File: src/libs/kata-types/src/config/hypervisor/mod.rs
Comment (pmores) [reply src/libs/kata-types/src/config/hypervisor/mod.rs:1058]: I understand the motivation behind this as those two keys only pertain to the remote hypervisor.  However it's rather inconsistent with how the whole config info is stored.  There are many other keys that only pertain to a subset of hypervisors, or even a single one, and they are still direct members of their config structures for better or worse.

Arguably these two keys should be members of perhaps `BootInfo` since they seem to be what's need to "boot up" the remote hypervisor.

File: src/runtime-rs/config/configuration-remote.toml.in
Comment (pmores) [inline src/runtime-rs/config/configuration-remote.toml.in:34]: It's good to have it documented here that the remote hypervisor doesn't support NVDIMM but it should also be enforced in `validate()`, right?

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:158]: Come to think of it, why are we even doing this in the first place?

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (pmores) [inline src/runtime-rs/crates/hypervisor/src/remote/inner.rs:168]: Another `new_ttrpc_client()`?  Is there something that I'm missing about this call?

File: src/runtime-rs/config/configuration-remote.toml.in
Comment (Chasing1020) [reply src/runtime-rs/config/configuration-remote.toml.in:34]: `cloud-hypervisor` and `qemu` also documented that they doesn't support NVDIMM devices, and I cannot find related check in `validate()`. Could you please give me some more details about which one I should check?

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:168]: You mean maybe I can make the `client` as a field inside the struct `RemoteInner`?

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:158]: Thank you for your comment! I will remove this check later.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (pmores) [inline src/runtime-rs/crates/hypervisor/src/remote/inner.rs:104]: Why don't we use hypervisor config to supply values for vcpus and memory?

Also, how about image path?

File: src/runtime-rs/config/configuration-remote.toml.in
Comment (pmores) [reply src/runtime-rs/config/configuration-remote.toml.in:34]: Hm, qemu supports nvdimm so it's no surprise you can't see a guard against it.  I don't know about clh, if it doesn't support nvdimm and doesn't check then it probably should.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:168]: Ah okay, I assumed that we have that already and overlooked that we don't - sorry, my bad!

With that being the case my point is rather moot, though I do wonder a bit what the advantage is of making a new connection every time we need to communicate.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:158]: Right, thanks!

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:168]: I just inspired by the func `openRemoteService` in golang, which opens a new connection everytime. But actually every rpc method in the ttrpc-client only takes `&self`, maybe it's a good idea to move the client to the `RemoteInner` to reduce the cost of making a new connection.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:104]: Thanks for your carefulness, I noticed that patch of enable image path annotation for remote hypervisor already merged 2 month ago, I will add it soon.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:168]: I'm aware that the go code works like this, too.

I mean, I don't have a strong opinion at this point.  If the hypervisor protocol is a datagram one then the difference between storing a client instance vs. recreating it every time probably doesn't matter at all.  If the protocol is a stream one then it'd probably be cleaner to store the client in `RemoteInner`.

Personally I'd probably prefer storing the client if there's no strong argument against, just because that's how it's usually done.  That's just personal preference though, feel free to ignore it.

Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:168]: Yes, I also think what you said is absolutely correct! I'll try to make some modifications.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (pmores) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs:104]: If you feel though that this is out of scope for this PR (which has "basic boilerplate" in its title after all, not "full implementation" :-)) just let me know.

File: src/runtime-rs/crates/hypervisor/src/remote/inner.rs
Comment (Chasing1020) [reply src/runtime-rs/crates/hypervisor/src/remote/inner.rs]: I renamed the method `new_ttrpc_client()` to `get_ttrpc_client()`, ensuring that the client is lazily created only once. Additionally, I added a field `Option<Client>` to `RemoteInner`, which requires a manual implementation of `Default` trait.

**Comment by justxuewei** [comment]
@pmores @gkurz @Apokleos Hey guys. This project is sponsored by OSPP, a program like Google Summer of Code. The project will be considered incomplete if the patch doesn't get merged before deadline on Oct 31.

Could we somehow make this get merged as long as the CI happy? After this merged, you still are able to leave comments on this PR, @Chasing1020 files another PR and issue to solve them.

Sounds good to all of you?

**Review by justxuewei — APPROVED** [review]
Approval for quick merging, but further improvement is needed.

**Comment by justxuewei** [comment]
Apart from the way mentioned above, we can merge this patch and then revert it immediately. We cheat the system of OSPP that the work has been done. Indeed, most of the work has been done, except for some nits. The patch will be improved in a new PR based on this one. The advantage is that the implementation of this functionality doesn't need to be laid out in separate patches.

Anyway, this patch **MUST BE** merged tomorrow.

Thanks!

**Review by pmores — APPROVED** [review]
@justxuewei Understood, thanks for letting us know, approving to enable merge.

Honestly I'm fairly happy with this PR already, personally I don't think I'll have any more comments anyway!  I'm fine with handling outstanding requests in a follow-up PR not to complicate things too much.

Thanks @Chasing1020 , excellent job!

**Comment by justxuewei** [comment]
Got two approvals, I'm going to merge this patch. Thanks again to all of you who participated in this patch. @Chasing1020 @Apokleos @pmores @gkurz

**Comment by Chasing1020** [comment]
Thank you all very much for your time and comments. Participating in our open-source community is a very interesting experience. I will continue to improve and perfect this feature later.

**Pull request merged**
